src/Compilers/Core/Portable/SourceFileResolver.cs

```diff
@@ -39,21 +48,62 @@ public SourceFileResolver(ImmutableArray<string> searchPaths, string baseDirecto
 
             _baseDirectory = baseDirectory;
             _searchPaths = searchPaths;
-        }
+            _pathMap = pathMap;
 
-        public string BaseDirectory
-        {
-            get { return _baseDirectory; }
+            // the keys in pathMap should not end with a path separator
+            if (!pathMap.IsDefaultOrEmpty) foreach (var kv in pathMap)
+            {
+                var key = kv.Key;
+                if (key == null || key.Length == 0)
+                {
+                    throw new ArgumentException("empty key", nameof(pathMap));
```

Comment:
The "empty" word should be localizable. Perhaps entire string should be put in resource, if we don't want the word "key" to be localized, a comment can be added to the resource string about that.